pkg/auth/spiffe_authorizer.go

```diff
@@ -0,0 +1,90 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/buildbarn/bb-storage/pkg/digest"
+	pb "github.com/buildbarn/bb-storage/pkg/proto/configuration/auth"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+// SpiffeAuthorizer  authorizes based on Spiffe-IDs
+// https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md
+//
+// see spiffe.proto for configuration options
+type SpiffeAuthorizer struct {
+	*pb.SpiffeAuthorizer
+}
+
+// NewSpiffeAuthorizer creates a new SpiffeAuthorizer
+func NewSpiffeAuthorizer(config *pb.AuthorizerConfiguration) Authorizer {
+	spifAuth := config.GetSpiffe()
+	if spifAuth == nil {
+		return &SpiffeAuthorizer{spifAuth}
+	}
+	return &SpiffeAuthorizer{}
+}
+
+// Authorize implements the authorizer inferface
+func (s *SpiffeAuthorizer) Authorize(ctx context.Context, instanceNames []digest.InstanceName) []error {
+	fillErrors := func(err error) []error {
+		errs := make([]error, len(instanceNames))
+		if err != nil {
+			for i := range errs {
+				errs[i] = err
+			}
+		}
+		return errs
+	}
+	// Extract client certificate chain from the connection.
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return fillErrors(status.Error(codes.Unauthenticated, "Connection was not established using gRPC"))
+	}
+	tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return fillErrors(status.Error(codes.Unauthenticated, "Connection was not established using TLS"))
+	}
+	certs := tlsInfo.State.PeerCertificates
+	if len(certs) == 0 {
+		return fillErrors(status.Error(codes.Unauthenticated, "Client provided no TLS client certificate"))
+	}
```

Comment:
I think what we should aim for is to keep the implementations of Authorizer decoupled from the network protocols used. For example, if we ever need to implement certain parts of Buildbarn using different protocols (REST calls? GraphQL? Thrift?), it would be nice if we could still use the existing authorization framework. This is the reason why Authenticator is placed in pkg/grpc and Authorizer in pkg/auth.

My recommendation would be to let TLSClientCertificateAuthenticator wrap `ctx` upon success using `context.WithValue()`, attaching some kind of custom struct to it that contains the X.509 certificate. This Authenticator could then extract it without depending on any gRPC code.